lib/ruby_smb/client.rb

```diff
@@ -320,6 +320,15 @@ def initialize(dispatcher, smb1: true, smb2: true, smb3: true, username:, passwo
       @open_files = {}
 
       @smb2_message_id = 0
+
+      # The dispatcher can override some instance variables
+      if @dispatcher.instance_variable_defined?("@override_client_instance_variables")
```

Comment:
This new instance variable may be reasonable to add to the client interface in this library. This would enable consumers of ruby_smb to documented knowledge of this variable and thereby know the option is available.

This also makes this code a bit simpler as:
```suggestion
      if @dispatcher.override_client_instance_variables
```

Reply:
After looking closer at the metasploit-framework side this is not needed at all the value injection should just respect the values passed in on the `:new` call.

Reply:
I agree that the override should be cleaned up a bit.

The `ShadowMitmDispatcher` must override key negotiation fields such as `session_id` which are not passable as arguments to the `:new` call.

Should I add all the required override fields as keyword arguments to the `:new` call instead of using the override?

Reply:
It looks like `session_id` is already accessible on the client object after creation, in fact all the instance variables in the client class look to have accessors that allow manipulation:
https://github.com/rapid7/ruby_smb/blob/1ccd35e65b4750102f74c606935d0f96ee602f8b/lib/ruby_smb/client.rb#L127-L130

Since fields are not consumed during initialization they could simply be set afterwards.
```Ruby
client = RubySMB::Client.new(dispatcher, smb1: false, smb2: true, smb3: false, always_encrypt: true, username: '', password: '')

# set any instance values not taken by initializer before consuming client:
client.session_id = custom_dispatcher.session_id

client.tree_connect(path)
```

Reply:
I have removed the custom override attribute.
The module now sets the fields after initialization.